### PR TITLE
Story Poster: Use dimensions array instead of size name

### LINKED ISSUE #12354: Editor: incorrect poster dimensions being returned
## Bug Description

Lucky found this obscure bug where the checklist in the editor shows a warning about incorrect poster dimensions, even though the poster definitely has the right size.

I tracked this down to some unexpected behavior in WordPress that boils down to honoring `$content_width`.

This only happens in the admin and not when accessing the REST API in an other way.

Here's how it works:

1. In the editor we preload REST API responses, such as the one for the current story (which is handled by `Stories_Controller`)
2. `\Google\Web_Stories\REST_API\Stories_Controller::get_story_poster()` calls `wp_get_attachment_image_src()`
3. `wp_get_attachment_image_src()` in turn calls `image_downsize`
4. `image_downsize` in turn calls `image_constrain_size_for_editor`
5. `image_constrain_size_for_editor` then reduces the width and height because we're in the admin (`is_admin()` is true because of the REST API preloading)

To work around this, I think we have to call `image_get_intermediate_size()` to get the actual size ourselves, and if that returns `false` (because the size does not exist), get the original size from  `wp_get_attachment_metadata()`.

But there might be other options.

**Notes:**

* This same issue theoretically exists in `\Google\Web_Stories\Model\Story::load_from_post()`
* We could just use the `Story` model in `Stories_Controller`, so we can delete the redundant `get_story_poster` method there

**Aside:**

I find it unexpected that `wp_get_attachment_image_src()` honors `$content_width` like this. Maybe worth opening a Trac ticket?

## Expected Behaviour

There should be no incorrect warning.

## Steps to Reproduce

1. Activate Twenty Twenty theme
7. Create a new story and add a few pages
8. Add a poster image by uploading a new image that needs to be cropped & crop it
9. Save the story
10. Reload the page
11. Notice the checklist warning about incorrect poster dimensions

## Screenshots

Screenshot 2022-09-21 at 17 41 01

## Additional Context

- Plugin Version:
- WordPress Version:
- Operating System:
- Browser:

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary
Another solution to #12354. If you pass an array of dimensions instead of passing size name it bypasses the content width check. [See](https://github.com/WordPress/wordpress-develop/blob/5eddc8f18c97d2eb909d0f70f4e20732398a7bda/src/wp-includes/media.php#L69-L71)

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

0. Activate Twenty Twenty theme
0. Create a new story and add a few pages
0. Add a poster image by uploading a new image that needs to be cropped & crop it
0. Save the story
0. Reload the page
0. View checklist, see no incorrect poster dimensions.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12354

### CODE REVIEW

**Comment by swissspidy** [comment]
Does it work?

As discussed, would be nice to have some sorts of tests verifying that this actually solves the reported bug. With this approach here it should be doable to that with PHP tests alone, no E2E tests needed (though that would be the last resort).

Alternatively, if we can't have tests, at the very least we should add comments explaining why we do it that way.

**Comment by spacedmonkey** [comment]
@swissspidy Unit tests added. Closing other PR in favour of this.

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php [hunk 1]
```diff
@@ -533,22 +533,71 @@ public function test_get_item_with_featured_image(): void {
...
 		$this->assertArrayHasKey( 'story_poster', $data );
+		$this->assertSame( Image_Sizes::POSTER_PORTRAIT_IMAGE_DIMENSIONS[0], $attachment_src[1] );
+		$this->assertSame( Image_Sizes::POSTER_PORTRAIT_IMAGE_DIMENSIONS[1], $attachment_src[2] );
 		$this->assertEqualSetsWithIndex(
 			[
 				'id'         => $attachment_id,
 				'url'        => $attachment_src[0],
-				'height'     => $attachment_src[1],
-				'width'      => $attachment_src[2],
+				'height'     => $attachment_src[2],
```
Comment (timarney) [inline tests/phpunit/integration/tests/REST_API/Stories_Controller.php]: Nit: can we flip these so it's width before height 
'width'     => $attachment_src[1],
'height'     => $attachment_src[2],

Little better readability IMO

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php [hunk 2]
```diff
@@ -533,22 +533,71 @@ public function test_get_item_with_featured_image(): void {
...
+		$data     = $response->get_data();
+
+		$this->assertArrayHasKey( 'story_poster', $data );
+		$this->assertSame( Image_Sizes::POSTER_PORTRAIT_IMAGE_DIMENSIONS[0], $attachment_src[1] );
+		$this->assertSame( Image_Sizes::POSTER_PORTRAIT_IMAGE_DIMENSIONS[1], $attachment_src[2] );
+		$this->assertEqualSetsWithIndex(
+			[
+				'id'         => $attachment_id,
+				'url'        => $attachment_src[0],
+				'height'     => $attachment_src[2],
```
Comment (timarney) [inline tests/phpunit/integration/tests/REST_API/Stories_Controller.php]: Same Nit as above 

```
'width' => $attachment_src[1],
'height' => $attachment_src[2],
```

Would be more readable

File: tests/phpunit/integration/tests/REST_API/Stories_Controller.php [hunk 3]
```diff
@@ -645,8 +695,8 @@ public function test_get_item_with_featured_image_and_hotlinked_poster(): void {
 			[
 				'id'         => $attachment_id,
 				'url'        => $attachment_src[0],
-				'height'     => $attachment_src[1],
-				'width'      => $attachment_src[2],
+				'height'     => $attachment_src[2],
```
Comment (timarney) [inline tests/phpunit/integration/tests/REST_API/Stories_Controller.php]: Same again here:

```
'width' => $attachment_src[1],
'height' => $attachment_src[2],
```

**Comment by swissspidy** [comment]
The tests don't work, as in, they demonstrate the bugfix. If I undo the changes to `includes/REST_API/Stories_Controller.php` and `includes/Model/Story.php` they still pass.

----

Note that for the QA instructions you can reuse the reproduction steps from the issue.

**Comment by spacedmonkey** [comment]
> The tests don't work, as in, they demonstrate the bugfix. If I undo the changes to includes/REST_API/Stories_Controller.php and includes/Model/Story.php they still pass.

The fix does work, I just used the wrong test data. I have updated the fixture data and tests pass and you can see them pass.

**Comment by swissspidy** [comment]
When I undo the changes to `includes/REST_API/Stories_Controller.php` I expect the `Stories_Controller` tests to fail, but they don't.

**Comment by swissspidy** [comment]
Also, running the tests creates a bunch of untracked files in the tests directory:

Screenshot 2022-10-13 at 08 38 11

**Review by swissspidy — CHANGES_REQUESTED** [review]
A couple of adjustments needed, as per my recent comments

**Comment by swissspidy** [comment]
QA instructions would be needed too here

**Comment by spacedmonkey** [comment]
> Also, running the tests creates a bunch of untracked files in the tests directory:
> 
> Screenshot 2022-10-13 at 08 38 11

I added delete attachment calls to all places where image sizes are generated.

**Comment by swissspidy** [comment]
> When I undo the changes to `includes/REST_API/Stories_Controller.php` I expect the `Stories_Controller` tests to fail, but they don't.

This still needs to be addressed AFAICT. After that this should be ready.

**Comment by spacedmonkey** [comment]
> This still needs to be addressed AFAICT. After that this should be ready.

I believe I fixed these issues with new test data.

**Comment by swissspidy** [comment]
Just verified that it's still an issue.

**Comment by spacedmonkey** [comment]
> Just verified that it's still an issue.
> 
> [3be82da](https://github.com/GoogleForCreators/web-stories-wp/commit/3be82daae767ee8bfb53c4776a33942d4435f4b0) was the only change since I flagged this.

Fix the unit test here. https://github.com/GoogleForCreators/web-stories-wp/pull/12418/commits/21de72a614073c7fd5a4c82ee3406c7b43602931

**Comment by swissspidy** [comment]
Excellent! 🎉

**Review by swissspidy — APPROVED** [review]

**Comment by kkalarickal** [comment]
Reproduced this problem (thanks to @LuckynaSan and @spacedmonkey for your help) by following these steps:

1. Switch to an older theme (I tried Twenty-Twenty, Twenty-Nineteen, Twenty-8teen, Twenty-Seventeen)
2. Create a sample web story and save to draft
3. Hit publish and update the poster image by selecting a larger image and cropping it
4. Exit from editor and reopen the same published story -- Now you will see the incorrect checklist error saying poster image size is not up to expected 640x853. This is in spite of the image being cropped properly.

Verified the fix by loading the `12418 branch` in the same wp environment with the older theme and repeating the steps above to confirm that the checklist does NOT report a non-existent problem.

**Pull request merged**